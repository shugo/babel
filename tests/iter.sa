class MAIN is
   repeat!(once n: INT) is
      i ::= 0;
      loop
         while!(i < n);
         yield;
         i := i + 1;
      end;
   end;

   upto!(once from, once to: INT): INT is
      i ::= from;
      loop
         while!(i <= to);
         yield i;
         i := i + 1;
      end;
   end;

   sum!(i: INT): INT is
      r ::= 0;
      loop
         r := r + i;
         yield r;
      end;
   end;

   product!(i: INT): INT is
      r ::= 1;
      loop
         r := r * i;
         yield r;
      end;
   end;

   main is
      loop
         repeat!(3);
         #OUT + "hello\n";
      end;
      sum: INT;
      product: INT;
      loop
         i ::= upto!(1, 10);
         #OUT + "i=" + i + "\n";
         sum := sum!(i);
         product := product!(i);
      end;
      #OUT + "sum=" + sum + "\n";
      #OUT + "product=" + product + "\n";
   end;
end;
