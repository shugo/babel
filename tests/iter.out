hello
hello
hello
i=1
i=2
i=3
i=4
i=5
i=6
i=7
i=8
i=9
i=10
sum=55
product=3628800
